# Problem 3

In [1]:
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


Epoch: 1/5.. Training loss: 0.486.. Test loss: 0.432.. Test Accuracy: 0.841
Epoch: 2/5.. Training loss: 0.373.. Test loss: 0.400.. Test Accuracy: 0.853
Epoch: 3/5.. Training loss: 0.336.. Test loss: 0.392.. Test Accuracy: 0.859
Epoch: 4/5.. Training loss: 0.314.. Test loss: 0.427.. Test Accuracy: 0.851
Epoch: 5/5.. Training loss: 0.299.. Test loss: 0.365.. Test Accuracy: 0.871


In [2]:
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    #optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


Epoch: 1/5.. Training loss: 2.315.. Test loss: 2.668.. Test Accuracy: 0.229
Epoch: 2/5.. Training loss: 2.715.. Test loss: 2.805.. Test Accuracy: 0.195
Epoch: 3/5.. Training loss: 2.899.. Test loss: 3.302.. Test Accuracy: 0.233
Epoch: 4/5.. Training loss: 4.155.. Test loss: 3.258.. Test Accuracy: 0.178
Epoch: 5/5.. Training loss: 2.956.. Test loss: 3.321.. Test Accuracy: 0.118


In lecture, we have learned that the place where the gradient resets lies in optimizer.zero_grad(). The following function will make the gradient to be zero. If the function is not called, the gradient will be accumulated throughtout the functions. Before making any changes, I have obtained the following result:
- Epoch: 1/5
  - Training loss: 0.486
  - Test loss: 0.432
  - Test Accuracy: 0.841
- Epoch: 2/5
  - Training loss: 0.373
  - Test loss: 0.400.. Test Accuracy: 0.853
- Epoch: 3/5
  - Training loss: 0.336
  - Test loss: 0.392
  - Test Accuracy: 0.859
- Epoch: 4/5
  - Training loss: 0.314
  - Test loss: 0.427
  - Test Accuracy: 0.851
- Epoch: 5/5 
  - Training loss: 0.299
  - Test loss: 0.365
  - Test Accuracy: 0.871

Running the code without optimizer.zero_grad, I have obtained the following result:
- Epoch: 1/5
  - Training loss: 2.315
  - Test loss: 2.668
  - Test Accuracy: 0.229
- Epoch: 2/5
  - Training loss: 2.715
  - Test loss: 2.805
  - Test Accuracy: 0.195
- Epoch: 3/5
  - Training loss: 2.899
  - Test loss: 3.302
  - Test Accuracy: 0.233
- Epoch: 4/5
  - Training loss: 4.155
  - Test loss: 3.258
  - Test Accuracy: 0.178
- Epoch: 5/5
  - Training loss: 2.956
  - Test loss: 3.321
  - Test Accuracy: 0.118

Comparing the two, one with the optimizer.zero_grad actually has performed much better. While one with optimizer.zero_grad has shown some type of pattern where increasing the epoch increase the test accuracy (for most case). On the other hand, one without the zero_grad appears to have decrease in test accuracy as the number of epoch increases.